proposals/0018/README.md

```diff
@@ -0,0 +1,191 @@
+---
+id: 0018
+title: Add Instance Object Definition
+status: ideation
+authors: Manny <mmendez@equinix.com>
+---
+
+## Summary
+
+The current hardware data model does not support any OS installation details.
+Installing an OS is one of the main focus areas of the Tinkerbell project at the moment.
+This exists in the legacy Packet Platform under "instance" in the hardware object and a similar concept is in [metal3].
+
+This would be used by Boots and Hegel (at least) to serve their needs.
+
+## Goals and not Goals
+
+We have a goal of keeping the Hardware data relatively static.
+We need a locaion to place instance/provision/installation-info that changes from provision to provision.
+It is not a goal to stick this info in `elastic_data`.
+
+## Content
+
+This proposal is about adding a storage concept for data that is not as permanent as Hardware, and that can persist across multiple Workloads.
+This would mimic the EquinixMetal life-cycle of a machine, provisioned and thus occupied until the customer deletes and the machine is deprovisioned.
+This life-cycle is also useful in non EM use cases as it allows a worflow to end and a future workflow to act upon the previous os intallation state.
+
+#### Hegel
+
+Hegel currently serves up EC2 style metadata by transforming the JSON blob in hardware.metadata to the compatible form.
+The JSON blob in Hardware.metadata is not specified in order to be flexible to non EquinixMetal operators, yet Hegel's transformation logic is fixed.
+So in reality this is just an undocumented JSON blob that is actually very useful to the project overall.
+
+#### Boots
+
+Boots is currently dynamically creating the iPXE script from data in Hardware.
+We generally want to keep the Hardware data static.
+But we would end up needing to modifying it to add ephemeral info, such as IPs (public/private v4 and public v6 as EquinixMetal does today), hostname, os installer to use...
+
+## APIs
+
+The API consists of defining a minimal and extensible Instance object as a field in Hardware, via Protobufs.
+We already have a protobuf file in tree ([packet]) that defines the non-tinkerbell Instance JSON object as Protobuf.
+This will be used as the starting point, moving fields that are unnecessary for Tinkerbell/Hegel/Boots into a meta field.
+If interest builds we can hoist meta values into the Instance protobuf definition.
+
+```protobuf
+syntax = "proto3";
+
+option go_package = "packet";
+
+package github.com.tinkerbell.tink.protos.packet;
+
+message Instance {
+  string id = 1;
+  string state = 2;
+  string hostname = 3;
+
+  // fork in the road! choose a path
+  // path 1Left
+  message CustomIPXE {
+    string url = 1;
+    bool always_pxe = 2;
+  }
+  message OperatingSystem {
+    string distro = 1;
+    string version = 2;
+    string tag = 3; // snapshot? revision? ...?
```

Comment:
Are these two fields going to be sufficient to enumerate the different types of versioning practices for various OSes? Or will it lead to trying to shoehorn everything into the model we provide?